Prototype implementation of activity timeline feature

GrimJu opened the pull request:
#293 Prototype implementation of activity timeline feature

As promised, a prototype implementation. I deviated from my original idea/design and used the material UI timeline for simplicity.

Feels quite hacky in some places, but I'm quite happy with the outcome. Especially the SQL part could definitly improved. Performance was very good on my 6000 activity database. Feel free to change/ improve stuff if you see it necessary.

**I'm not sure if the database migration is correct. I've never heard of knex before and just pasted my function creation script in there. I did not test it.**

This is what it looks like:
![image](https://github.com/user-attachments/assets/0aa5b681-5967-4b0f-a6a9-cd339cfe9a59)

CyferShepard commented:
Hey @GrimJu , this looks great. I have a few ideas to make it more user friendly
1) Since the Item Id is being passed back in the response, maybe add a hyperlink around the cards so that we can directly navigate to the item shown in the timeline
2) Since we have to already select a user to render the graph, why not move this view to a new tab under user? Like her for instance 
![image](https://github.com/user-attachments/assets/5951e732-58fe-4e66-835c-ad2d89892053)

If youd still like this page to also exist outside of the users page to reduce clicks needed to navigate to it, i suggest adding under another screen, like statistics or maybe under activity, and give it the tab selector to switch between the 2. If this sounds too hard to implement let me know and i can assist where i can.

3) I see an episode count is also in the response, but not displayed, i feel like this would be a great addition on the ui. So something like Season 1 - 5 Episodes

Thanks again for this cool feature

GrimJu commented:
> 1. Since the Item Id is being passed back in the response, maybe add a hyperlink around the cards so that we can directly navigate to the item shown in the timeline

Definitly

> 2. Since we have to already select a user to render the graph, why not move this view to a new tab under user? Like here for instance

Can also do that, I would also add it to activities with a new tab as you said

> 3. I see an episode count is also in the response, but not displayed, i feel like this would be a great addition on the ui. So something like Season 1 - 5 Episodes

I would like that in principle, but at least on my database this would be problematic. Not all episodes are recorded in my database so showing the number of episodes would be missleading. I originally tried to only show completed seasons but ran into the same problem. I'm not sure on how to continue this, of course the correct behavior would be that every activity counts but that isn't the case for me. What would be your approach for this? 

I would also filter out all activities under a certain time threshhold.

GrimJu commented:
Implemented 1 and 3. 

For 2 I added the tab in the user-info. But I did keep the timeline's unique screen. I'm not sure on how adding it to another screen would improve the UX. 

I would ask you to reconsider not giving it its own screen, otherwise please help me figure out a concrete concept of where to add it and in what way.

I don't belive it fits under statistics as it's not number based and also the statistics tab is global for all users. I don't think adding one element to the page that's user specific while all others are server-wide is the right call.

As for the activity screen. I couldn't figure out a way in which all the necessary control elements on the top of the page look harmonious. Again if you have a concept for that, feel free to share it and I will implement it.

CyferShepard commented:
Hey @GrimJu , Yea I also wasn't so sure about making it global, I agree its perfect as a per user screen instead. In regards to the time threshold. I dont think 15 minutes is appropriate, sometimes you can have short films or even music, and those can be under 15 minutes, or even under 5. Im not sure what a sane value would be aswell. Il check this out in the morning and provide more feedback

GrimJu commented:
I realized it's probably better to use jf_libraries.CollectionType as the "MediaType" instead of calculating it. On my db a lot of activities seem to not have that filled correctly. But I think I can assume that that's because of bugs that have been / will be fixed and I should not bother trying to work around that. In that case I'd use a minimum of 15 seconds for music, 5 minutes for movies and 15 minutes for shows.

GrimJu commented:
No problem on the delay, I was sick myself and wouldn't have done any work anyway. 

In my (admitedly biased) opinion, I believe having a standalone menu item / page for the timeline is fine. Of course jellystat is your project and I don't want to (or can) convince you of anything you don't want to do. I trust your judgement in this case.

But in case you decide that the activity timeline should not have its own menu item / page, I ask of you that you provide me with a plan on where to put it and how that would look like from a UI/UX perspective (either a description or a mockup). I already tried putting it under activity and could not come up with a way that blends with the current design language properly.

I also ask that you decide on an algortithm to filter out "too short" activities for the timeline.  for my usecase doing what I described in https://github.com/CyferShepard/Jellystat/pull/297#issuecomment-2586756933 would suffice, but I admit that I have a very single-minded view of this feature and would like some more input from you.

Maybe there's a way to calculate a the time spent as a percentage and then use that, but I couldn't find an easy way to do that with the available data. As I don't know the datamodel of both jellystat and jellyfin very well I would ask you to investigate that.

CyferShepard commented:
Hey, in regards to collection type, thats a field provided by Jellyfin, im assuming all of your fields are defaulting to mixed, thats a fallback on myside which means jellyfin isnt providing that field in the response, il need to investigate that as i dont have that issue. 

For the playback duration, i think 15 seconds should suffice, or 30 seconds but that would need testing to make sure theres no false positives when filtering data. Currently time spent as a percentage is calculated in jellystat as PlaybackDuration / (RuntimeTicks from the jf_library_items or jf_library_episodes table depending on the playback type), you can reference ActivityMonitor.js line 189 as an example.

In terms of the UI, i had something similar to the settings page in mind, its actually a collection of different views and the tab selector cycles between them, let me know what you think. If that still feels weird in terms of usability, il think of something else. If thats fine i can implement it to maintain the UI aesthetic

GrimJu commented:
> Hey, in regards to collection type, thats a field provided by Jellyfin, im assuming all of your fields are defaulting to mixed, thats a fallback on myside which means jellyfin isnt providing that field in the response, il need to investigate that as i dont have that issue.

That's not the issue. For some reason a lot of old music activities had my tvshows library linked as their library. I don't know if that still occurs or not as I turned of monitoring of music on my instance a few months ago.

> For the playback duration, i think 15 seconds should suffice, or 30 seconds but that would need testing to make sure theres no false positives when filtering data. Currently time spent as a percentage is calculated in jellystat as PlaybackDuration / (RuntimeTicks from the jf_library_items or jf_library_episodes table depending on the playback type), you can reference ActivityMonitor.js line 189 as an example.
> 

I just changed it to 30 seconds for now. My ideal scenario would be to mimic the behavior as defined in the /playback/resume section of jellyfin
![image](https://github.com/user-attachments/assets/502709c8-749c-4056-977a-17a4844179ea)
But I will leave implementing that to you/someone else if the need arises in the future.

> In terms of the UI, i had something similar to the settings page in mind, its actually a collection of different views and the tab selector cycles between them, let me know what you think. If that still feels weird in terms of usability, il think of something else. If thats fine i can implement it to maintain the UI aesthetic

I think it's fine, but I don't want to mess up your naming / code-style in existing components. Thus I would leave implementing that to you. I removed the sidebar link in my latest commit.

CyferShepard commented:
Will approve this and add the changes to the activity tab later. Im not sure if you forgot to check in the changes for the nav bar changes. I still see it in the change request.

GrimJu commented:
> Will approve this and add the changes to the activity tab later. Im not sure if you forgot to check in the changes for the nav bar changes. I still see it in the change request.

You're right I'm sorry, fixed it.

GrimJu commented:
Thanks for the fast merge, when is 1.1.3 slated for release?

CyferShepard commented:
So 1.1.3 is slated for this weekend or next weekend, im rewriting how the tasks works to fix alot of issues and add options to actually stop a task. The tasks themselves should change, just how they're implemented on a schedule and how they're being executed

GrimJu commented:
Hey, I'm sorry for reopening this discussion after you merged the PR. But I just now realized a kinda big flaw in how I built the SQL side of this. The way I've done the function completely erases re-watches of media.  After some chatgpt magic I came up with an updated function that separates groups based on if their ActivityDateInserted is at least one month apart (please choose whatever value you want for this). It seems to work in the (admittedly limited) testing I've done.

I ask of you to verify / test this new function and decide whether you want to use it or stick to the old one. In case you approve it I would create another PR for that. 

```sql
CREATE OR REPLACE FUNCTION fs_get_user_activity(
    user_id TEXT,
    library_ids TEXT[]
)
RETURNS TABLE (
    "UserName" TEXT,
    "Title" TEXT,
    "EpisodeCount" BIGINT, 
    "FirstActivityDate" TIMESTAMPTZ,
    "LastActivityDate" TIMESTAMPTZ,
    "TotalPlaybackDuration" BIGINT,
    "SeasonName" TEXT,
    "MediaType" TEXT,
    "NowPlayingItemId" TEXT
) AS $$
BEGIN
RETURN QUERY
WITH NumberedEntries AS (
    SELECT
        jp."UserName" AS "UserNameCol",
        CASE
            WHEN jp."SeriesName" IS NOT NULL THEN jp."SeriesName"
            ELSE jp."NowPlayingItemName"
        END AS "TitleCol",
        jp."EpisodeId" AS "EpisodeIdCol",
        jp."ActivityDateInserted" AS "ActivityDateInsertedCol",
        jp."PlaybackDuration" AS "PlaybackDurationCol",
        ls."Name" AS "SeasonNameCol",
        jl."CollectionType" AS "MediaTypeCol",
        jp."NowPlayingItemId" AS "NowPlayingItemIdCol",
        ROW_NUMBER() OVER (PARTITION BY jp."UserName" ORDER BY jp."ActivityDateInserted") AS rn
    FROM
        public.jf_playback_activity AS jp
    JOIN 
        public.jf_library_items AS jli ON jp."NowPlayingItemId" = jli."Id"
    JOIN 
        public.jf_libraries AS jl ON jli."ParentId" = jl."Id"
    LEFT JOIN 
        public.jf_library_seasons AS ls ON jp."SeasonId" = ls."Id"
    WHERE
        jp."UserId" = user_id
        AND jl."Id" = ANY(library_ids)
),
DateDifferences AS (
    SELECT
        "UserNameCol",
        "TitleCol",
        "EpisodeIdCol",
        "ActivityDateInsertedCol",
        "PlaybackDurationCol",
        "SeasonNameCol",
        "MediaTypeCol",
        "NowPlayingItemIdCol",
        rn,
        LAG("ActivityDateInsertedCol") OVER (PARTITION BY "UserNameCol", "TitleCol" ORDER BY "ActivityDateInsertedCol") AS prev_date
    FROM
        NumberedEntries
),
GroupedEntries AS (
    SELECT
        "UserNameCol",
        "TitleCol",
        "EpisodeIdCol",
        "ActivityDateInsertedCol",
        "PlaybackDurationCol",
        "SeasonNameCol",
        "MediaTypeCol",
        "NowPlayingItemIdCol",
        rn,
        prev_date,
        CASE
            WHEN prev_date IS NULL OR "ActivityDateInsertedCol" > prev_date + INTERVAL '1 month' THEN 1 -- Pick whatever interval you want here, I'm biased as I don't monitor music / never intended this feature to be used for music
            ELSE 0
        END AS new_group
    FROM
        DateDifferences
),
FinalGroups AS (
    SELECT
        "UserNameCol",
        "TitleCol",
        "EpisodeIdCol",
        "ActivityDateInsertedCol",
        "PlaybackDurationCol",
        "SeasonNameCol",
        "MediaTypeCol",
        "NowPlayingItemIdCol",
        rn,
        SUM(new_group) OVER (PARTITION BY "UserNameCol", "TitleCol" ORDER BY "ActivityDateInsertedCol") AS grp
    FROM
        GroupedEntries
)
SELECT
    "UserNameCol" AS "UserName",
    "TitleCol" AS "Title",
    COUNT(DISTINCT "EpisodeIdCol") AS "EpisodeCount",
    MIN("ActivityDateInsertedCol") AS "FirstActivityDate",
    MAX("ActivityDateInsertedCol") AS "LastActivityDate",
    SUM("PlaybackDurationCol")::bigint AS "TotalPlaybackDuration",
    "SeasonNameCol" AS "SeasonName",
    MAX("MediaTypeCol") AS "MediaType",
    "NowPlayingItemIdCol" AS "NowPlayingItemId"
FROM
    FinalGroups
GROUP BY
    "UserNameCol",
    "TitleCol",
    "SeasonNameCol",
    "NowPlayingItemIdCol",
    grp
HAVING
    NOT (MAX("MediaTypeCol") = 'Shows' AND "SeasonNameCol" IS NULL)
    AND SUM("PlaybackDurationCol") >= 30
ORDER BY
    MAX("ActivityDateInsertedCol") DESC;
END;
$$ LANGUAGE plpgsql;
```

GrimJu commented:
Hey @CyferShepard sorry for pinging you on this but I would really like your input on this as I'd hate for the feature to get released with such a glaring issue.

CyferShepard commented:
Hey @GrimJu , sorry i got caught up with testing other features i forgot about this, il test it this weekend and reply back here

CyferShepard commented:
Hey @GrimJu , im finally free to take a look at this code. Just wanted some clarification. From my understanding, if a show is re-watched, the timeline doesnt get updated, and only shows the first watch instance? And that would mean your fix is to allow rewatches to be taken into account and moved further up the timeline? or does it just add a new instance on the timeline? eg show A watch on 1st of Jan, Show B watched on 10th of Jan, Show A re-watched on 12th of Jan. This would result in a timeline of Show A----Show B ---- Show A

GrimJu commented:
Not quite but close:

> if a show is re-watched, the timeline doesnt get updated, and only shows the first watch instance?

It only shows the last (newest) instance due to the group by clause not including any timing information.

> And that would mean your fix is to allow rewatches to be taken into account and moved further up the timeline? or does it just add a new instance on the timeline? eg show A watch on 1st of Jan, Show B watched on 10th of Jan, Show A re-watched on 12th of Jan. This would result in a timeline of Show A----Show B ---- Show A

The second interpretation is correct, due to the updated funtion only grouping  entries that are at least (whatever interval you feel is best) apart it would show the timeline as Show A----Show B ---- Show A.

CyferShepard commented:
Cool, i tested it out, and it seems to work. I feel 30 seconds is abut too high, is it fine if we can lower this value abit? maybe like 20 seconds?

GrimJu commented:
It's your project, choose whatever number you fell like, maybe different numbers for different mediatypes. You should also think about a timeframe from when two activities count as a rewatch, I used 30 days for testing purposes and I think it's a good starting point for shows and movies but I'd still want feedback on that from you. Maybe even make that user-configurable

CyferShepard commented:
So i think 20 seconds min play duration, and 30 days seems fine, im not sure how youd make that configurable, of if we really need to anyway

GrimJu commented:
One could make it configurable by adding a parameter for the interval and save that in the app_config.settings json. Fine if you think it's unnecessary but I could add it should the need ever arise. 

As for the changes to the script. Will you commit that yourself or should I create another PR?

CyferShepard commented:
Il make the change myself, plan to release tomorrow
